binex/src/decoder.rs

```diff
@@ -0,0 +1,64 @@
+use std::io::Read;
+
+use log::{debug, error};
+
+use crate::{constants::Constants, message::Message, utils::Utils, Error};
+
+#[derive(Debug, Copy, Clone, PartialEq, Default)]
+pub enum State {
+    /// Searching for SYNC byte, defining a start of stream
+    #[default]
+    Synchronizing = 0,
+    /// Message ID byte follows SYNC byte and is BNXI encoded
+    MID = 1,
+    /// Message length follows MID bytes and is BNXI encoded
+    MLength = 2,
+    /// Message content follows MLength bytes
+    Message = 3,
+}
+
+/// [BINEX] Stream Decoder
+pub struct Decoder<R: Read> {
+    /// [R]
+    reader: R,
+    /// Buffer read pointer
+    ptr: usize,
+    /// Internal buffer
+    buffer: Vec<u8>,
+    /// Current Message being decoded
+    msg: Message,
+    /// Internal [State]
+    state: State,
+    /// Minimal number of bytes for current [State]
+    next_read: usize,
+}
+
+impl<R: Read> Decoder<R> {
+    pub fn new(reader: R) -> Self {
+        Self {
+            ptr: 0,
+            reader,
+            next_read: 128,
+            state: State::default(),
+            msg: Default::default(),
+            buffer: [0; 128].to_vec(),
+        }
+    }
+}
+
+impl<R: Read> Iterator for Decoder<R> {
+    type Item = Result<Message, Error>;
+    /// Parse next message in stream
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.reader.read(&mut self.buffer) {
+            Ok(size) => {},
+            Err(e) => {
```

Comment:
Shouldn't this propagate the error (`self.reader.read(&mut self.buffer)?`)? When the end of the stream is reached, `read` returns 0, not an error.

Reply:
Yes it needs to propage at least the EOS error (and possibly other major errors), because that's the only way for the user to determine that the EOS is reached.

I'm currently working on the actual message decoding and am not focused on the Decoder (final layer)

Reply:
It's starting to work, I'm able to fill the buffer (I assigned an arbitrary size of 1024), 128 should be minimum.
For some reason, I can only get the I/O read interface to work, if I allocate my Vector with `vec = [0; 1024]`, which seems to _actually_ fill the buffer. At first I used `Vec::with_capacity(1024)`, which does not seem to fill the buffer, `.read()` then presumes the buffer is actually empty :/